tests/js/core/widget/core/Drawer/Drawer.js

```diff
@@ -0,0 +1,686 @@
+				direction: "right"
+			};
+			widget.options = {
+				enable: true,
+				position: "left"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from right");
+
+			// case left
+			event.detail = {
+				direction: "left"
+			};
+			widget.options = {
+				enable: true,
+				position: "right"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from left");
+
+			// case up
+			event.detail = {
+				direction: "up"
+			};
+			widget.options = {
+				enable: true,
+				position: "down"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from up");
+
+			// case down
+			event.detail = {
+				direction: "down"
+			};
+			widget.options = {
+				enable: true,
+				position: "up"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from down");
+
+			// reset event object
+			event = {};
+
+			// direction provided by event type
+			// case right
+			event.type = "swiperight"
+			widget.options = {
+				enable: true,
+				position: "left"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from right");
+
+			// case left
+			event.type = "swipeleft";
+			widget.options = {
+				enable: true,
+				position: "right"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from left");
+
+			// case up
+			event.type = "swipeup";
+			widget.options = {
+				enable: true,
+				position: "down"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from up");
+
+			// case down
+			event.type = "swipedown";
+			widget.options = {
+				enable: true,
+				position: "up"
+			};
+			widget._isDrag = true;
+			widget._onSwipe(event);
+			assert.ok(!widget._isDrag, "widget swipe from down");
+		});
+
+		test("Test of _translate drawer", 8, function (assert) {
+			var widget = new Drawer(),
+				element = element = document.getElementById("drawer-1");
+
+			widget.element = element;
+			widget._state = "not settling";
+			widget.options = {
+				overlay: true
+			}
+
+			helpers.stub(widget, "_setOverlay", function (x, y) {
+				assert.ok(true, "_setOverlay has been called");
+				assert.equal(x, 10, "x provided as 10");
+				assert.equal(y, 20, "y provided as 20");
+			});
+			helpers.stub(widget, "_onTransitionEnd", function () {
+				assert.ok(true, "_onTransitionEnd has been called");
+			});
+
+			// case 1
+			widget._translate(10, 20, 1000);
+
+			assert.equal(widget._state, "sliding", "widget _state set to 'sliding'");
+			// this test depends to web engine because: transform, -webkit-transform
+			//assert.ok(!!element.style.transition, "transform style has been set'");
+
+			// case 2 (duration is 0)
+			widget._state = "not settling";
+			widget._translate(10, 20, 0);
```

Comment:
some questions
1. no need to check widget state?
2. shouldn't we check _translatedX here?

Reply:
1. I think we don't need check widget state. It has already been checked in previous case.
I can add this check if you wish.
2. We have to call `_translate` with duration `0` to check if the method `_onTransitionEnd` will be call.